File: binary_tree/404_sum_of_left_leaves.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sumOfLeftLeaves(root: TreeNode | null): number {

    function dfs(node, left){

        if(!node){
            return 0;
        }
        if(!node.left && !node.right && left){
            return node.val;
        }
        return dfs(node.left, true) + dfs(node.right, false)
    }

    return dfs(root.left, true) + dfs(root.right, false)

};

File: binary_tree/404_sum_of_left_leaves.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {

        pub fn dfs(node: Option<Rc<RefCell<TreeNode>>>, left: bool) -> i32 {
            match node {
                None => 0,
                Some(node) => {
                    let borrow_node = node.borrow();
                    if left && borrow_node.left.is_none() && borrow_node.right.is_none() {
                        borrow_node.val
                    } else {
                        dfs(borrow_node.left.clone(), true) + dfs(borrow_node.right.clone(), false)
                    }
                }
            }
        }

        if let Some(our_root) = root {
            let borrow_root = our_root.borrow();
            dfs(borrow_root.left.clone(), true) + dfs(borrow_root.right.clone(), false)
        } else {
            0
        }
    }
}